### PR TITLE
Fixes rare RingBuffer initialization issue (#4649)

### PR DESCRIPTION
**Pull request by nyanzebra**
cherry-pick of https://github.com/Azure/iotedge/pull/4649

### CODE REVIEW

**Comment by varunpuranik** [comment]
@nyanzebra - how did we find this issue? How can we make sure we have covered all cases?
Also - can you please update the title (this is not just a test, but a code fix), and description to the fix as well?

**Comment by nyanzebra** [comment]
> 
> 
> @nyanzebra - how did we find this issue? How can we make sure we have covered all cases?
> Also - can you please update the title (this is not just a test, but a code fix), and description to the fix as well?

Yes, updated the title. I still try to test this code when I get chances to do so. That is how I found this. I think due to the limited number of possibilities on initialization, that I actually can accurately write out all possible cases that can happen. Of which I think there are 4 types (can vary these tests below based on full and not full buffer):
1. Re-init after no work
2. Re-init after writing and/or reading where write pointer is after the read pointer
3. Re-init after writing and reading where write pointer is behind read pointer (wrap-around)
4. Re-init after writing and/or reading where write pointer is at same position as read pointer (should be able to test both 0 reads and some reads and verify the flag for wrap around is set)

There is of course the case of just init with no previous data, but the above are the interesting ones. So, in thinking about it, these are the types of cases I should make sure exist.

This is a cherry-pick into 1.2, so let me see if there are additional tests and maybe bugs that need to be fixed by going through the cases above on initialization.

File: mqtt/mqtt-bridge/src/persist/waking_state/ring_buffer/mod.rs [hunk 1]
```diff
@@ -435,19 +435,24 @@ fn find_pointers_and_order_post_crash(file: &mut File, max_file_size: u64) -> Ri
 
         // Check next block
         write_updates += 1;
-        order += 1; // We want to start with the next order number.
         start = end + data_size;
         end = start + block_size;
 
         // Found the last write, can stop updating write.
-        if inner.order() < order {
+        if should_update_write && inner.order() < order {
```
Comment (varunpuranik) [inline mqtt/mqtt-bridge/src/persist/waking_state/ring_buffer/mod.rs:442]: What does this flag indicate?

**Comment by varunpuranik** [comment]
Thanks for the details, @nyanzebra. 
For re-init, we need 3 pieces of info - 
- read ptr
- write ptr
- flag to indicate if read_index is ahead of write_index (empty vs full case)
Do we still need to take care of all these cases? If so, why?

**Review by dmolokanov — APPROVED** [review]

**Comment by nyanzebra** [comment]
> 
> 
> Thanks for the details, @nyanzebra.
> For re-init, we need 3 pieces of info -
> 
>     * read ptr
> 
>     * write ptr
> 
>     * flag to indicate if read_index is ahead of write_index (empty vs full case)
>       Do we still need to take care of all these cases? If so, why?

Correct we need all the data found in the metadata struct. 

We have test cases on these from double checking the kinds of tests that exist, minus the one test case this adds. These are the 'it_inits_...' tests. I am re-evaluating these tests and also seeing if there is a way to simplify the tests to make them easier to evaluate for correctness. I am trying to fuzz things and walk through the code multiple time on the side to make this more robust, and is probably why I am finding some issues.

**Pull request merged**